contracts/accessless/OlympusChainlinkBasedSwapper.sol

```diff
@@ -0,0 +1,215 @@
+        router.swapExactTokensForTokensSupportingFeeOnTransferTokens(
+            params.amount,
+            _calcAmountMin(inputToken, outputToken, denomination, params.slippage, params.amount),
+            params.path,
+            params.recipient,
+            (params.deadline == 0) ? block.timestamp : params.deadline
+        );
+
+        // cleanup
+        _clean(inputToken);
+    }
+
+    function _exact(V3Params calldata params, bool isIn) internal returns (uint256 amount) {
+        // reads + calcs
+        address inputToken;
+        address outputToken;
+
+        if (isIn) {
+            inputToken = params.path[0];
+            outputToken = params.path[params.path.length - 1];
+        } else {
+            inputToken = params.path[params.path.length - 1];
+            outputToken = params.path[0];
+        }
+
+        address denomination = (params.denomination == address(0)) ? Denominations.USD : params.denomination;
+
+        // checks
+        if (params.path.length - 1 != params.fees.length) revert OlympusChainlinkBasedSwapper_WrongV3PathLength();
+
+        // interactions + checks
+
+        bool check = IERC20(inputToken).transferFrom(msg.sender, address(this), params.amount);
+
+        if (!check) revert OlympusChainlinkBasedSwapper_TransferFromFailed();
+
+        check = IERC20(inputToken).approve(address(v3SwapRouter), params.amount);
+
+        if (!check) revert OlympusChainlinkBasedSwapper_ApproveFailed();
+
+        // big interaction
+        if (isIn) {
+            amount = v3SwapRouter.exactInput(
+                ISwapRouter.ExactInputParams({
+                    path: _encode(params.fees.length, params.path, params.fees),
+                    recipient: params.recipient,
+                    deadline: (params.deadline == 0) ? block.timestamp : params.deadline,
+                    amountIn: params.amount,
+                    amountOutMinimum: _calcAmountMin(
+                        inputToken,
+                        outputToken,
+                        denomination,
+                        params.slippage,
+                        params.amount
+                    )
+                })
+            );
+        } else {
+            amount = v3SwapRouter.exactOutput(
+                ISwapRouter.ExactOutputParams({
+                    path: _encode(params.fees.length, params.path, params.fees),
+                    recipient: params.recipient,
+                    deadline: (params.deadline == 0) ? block.timestamp : params.deadline,
+                    amountOut: _calcAmountMin(inputToken, outputToken, denomination, params.slippage, params.amount),
+                    amountInMaximum: params.amount
+                })
+            );
+        }
+
+        // cleanup
+        _clean(inputToken);
+    }
+
+    // @notice
+    function _clean(address inputToken) internal {
+        bool check = IERC20(inputToken).transfer(msg.sender, IERC20(inputToken).balanceOf(address(this)));
```

Comment:
are there edge cases where transferring the full balance is not the desired behavior?

Reply:
Tokens should never be stuck in the contract, so always transfer out. It should never happen that your tokens get locked inside it.